crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -535,6 +537,83 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
+        to_upgrade.each do |element|
+          # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
+          if ServiceObject.is_cluster?(element)
+            cluster = ServiceObject.cluster_name element
+            cluster_env = "pacemaker-config-#{cluster}"
+            founder = ::Node.find(
+              "pacemaker_founder:true AND " \
+              "pacemaker_config_environment:#{cluster_env}"
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
+            # FIXME: could we do it in paralel?
+            upgrade_one_node element
+          end
+        end
+
+        # and handle those as part of the compute node upgrade
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
+        to_upgrade = []
+
+        # First find out the nodes with compute role, for later checks
+        compute_nodes = {}
+        nova_prop = proposals["nova"][:prop] || nil
+        unless nova_prop.nil?
+          elements = nova_prop["deployment"]["nova"]["elements"]
+          elements.each do |role, nodes|
+            next unless role.start_with?("nova-compute")
+            nodes.each { |node| compute_nodes[node] = 1 }
+          end
+        end
+
+        # For each active proposal go through the roles in element order
+        proposals.each do |name, propmap|
+          proposal = propmap[:prop]
+          elements = proposal["deployment"][name]["elements"]
+          proposal["deployment"][name]["element_order"].flatten.each do |el|
+            # skip nova compute nodes, they will be upgraded separately
+            next if el.start_with?("nova-compute") || elements[el].nil?
+            # skip some roles if they are assigned to compute nodes
+            if ["cinder-volume", "ceilometer-agent", "swift-storage"].include? el
+              elements[el].each do |node|
```

Comment:
I think this won't work if e.g. ```cinder-volume``` is assigned to a cluster. In that case ```elements[el]``` will contain ```cluster:<name>```.  So it's probably better to expand the elements here. ```ServiceObject.expand_nodes_for_all``` should be able to do that.